src/java_bytecode/java_pointer_casts.cpp

```diff
@@ -0,0 +1,160 @@
+/*******************************************************************\
+
+Module: JAVA Pointer Casts
+
+Author: Daniel Kroening, kroening@kroening.com
+
+\*******************************************************************/
+
+#include <util/std_expr.h>
+#include <util/std_types.h>
+#include <util/namespace.h>
+
+#include "java_pointer_casts.h"
+
+/*******************************************************************\
+
+Function: clean_deref
+
+  Inputs: pointer
+
+ Outputs: dereferenced pointer
+
+ Purpose: dereference pointer expression
+
+\*******************************************************************/
+
+exprt clean_deref(const exprt &ptr)
+{
+  return ptr.id()==ID_address_of
+             ? ptr.op0()
+             : dereference_exprt(ptr, ptr.type().subtype());
+}
+
+/*******************************************************************\
+
+Function: find_superclass_with_type
+
+  Inputs: pointer
+          target type to search
+
+ Outputs: true iff a super class with target type is found
+
+ Purpose:
+
+\*******************************************************************/
+
+
+bool find_superclass_with_type(
+  exprt &ptr,
+  const typet &target_type,
+  const namespacet &ns)
+{
+  while(true)
+  {
+    const typet ptr_base=ns.follow(ptr.type().subtype());
```

Comment:
As this function isn't just static, more defensive programming is required here. You should guard against non-pointer types being passed in.